File: sdk-js/src/verify.js
import { canonicalize } from "./canonicalize.js";
import { base64UrlDecode, verifyCanonicalPayload } from "./crypto.js";
import { PayloadValidationError, SignatureVerificationError } from "./errors.js";
import { resolveTrustAnchorPublicKey } from "./trust-registry.js";
import { decodeTransportPayload } from "./qr.js";

function validateRequiredString(value, name) {
  if (typeof value !== "string" || value.length === 0) {
    throw new PayloadValidationError(`Missing or invalid ${name}`);
  }
}

function validateDocumentPaymentField(document, fieldName) {
  const value = document[fieldName];
  if (value === undefined || value === null || value === "") {
    return false;
  }
  return true;
}

function compareExpectedValue(actual, expected) {
  if (expected === undefined) {
    return true;
  }

  if (typeof actual === "number" || typeof expected === "number") {
    return Number(actual) === Number(expected);
  }

  return String(actual) === String(expected);
}

function validatePayload(payload) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new PayloadValidationError("Payload must be a JSON object");
  }

  validateRequiredString(payload.version, "version");
  validateRequiredString(payload.intent, "intent");
  validateRequiredString(payload.issued_at, "issued_at");
  validateRequiredString(payload.expires_at, "expires_at");
  validateRequiredString(payload.nonce, "nonce");
  validateRequiredString(payload.alg, "alg");
  validateRequiredString(payload.sig, "sig");

  if (payload.version !== "1") {
    throw new PayloadValidationError(`Unsupported version: ${payload.version}`);
  }

  if (!payload.issuer || typeof payload.issuer !== "object" || Array.isArray(payload.issuer)) {
    throw new PayloadValidationError("Missing or invalid issuer object");
  }

  validateRequiredString(payload.issuer.issuer_id, "issuer.issuer_id");
  validateRequiredString(payload.issuer.display_name, "issuer.display_name");
  validateRequiredString(payload.issuer.trust_anchor_id, "issuer.trust_anchor_id");

  if (!payload.document || typeof payload.document !== "object" || Array.isArray(payload.document)) {
    throw new PayloadValidationError("Missing or invalid document object");
  }

  validateRequiredString(payload.document.document_id, "document.document_id");

  return payload;
}

function validatePaymentProfilePayload(payload) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new PayloadValidationError("Payload must be a JSON object");
  }

  validateRequiredString(payload.version, "version");
  validateRequiredString(payload.intent, "intent");
  validateRequiredString(payload.issued_at, "issued_at");
  validateRequiredString(payload.expires_at, "expires_at");
  validateRequiredString(payload.nonce, "nonce");
  validateRequiredString(payload.alg, "alg");

  if (payload.version !== "1") {
    throw new PayloadValidationError(`Unsupported version: ${payload.version}`);
  }

  if (!payload.issuer || typeof payload.issuer !== "object" || Array.isArray(payload.issuer)) {
    throw new PayloadValidationError("Missing or invalid issuer object");
  }

  validateRequiredString(payload.issuer.issuer_id, "issuer.issuer_id");
  validateRequiredString(payload.issuer.display_name, "issuer.display_name");
  validateRequiredString(payload.issuer.trust_anchor_id, "issuer.trust_anchor_id");

  if (!payload.document || typeof payload.document !== "object" || Array.isArray(payload.document)) {
    throw new PayloadValidationError("Missing or invalid document object");
  }

  validateRequiredString(payload.document.document_id, "document.document_id");

  return payload;
}

function stripSignature(payload) {
  const { sig, ...body } = payload;
  return body;
}

function verifyTimeWindow(payload, now) {
  const issuedAt = Date.parse(payload.issued_at);
  const expiresAt = Date.parse(payload.expires_at);
  const currentTime = now instanceof Date ? now.getTime() : Date.parse(now);

  if (Number.isNaN(issuedAt) || Number.isNaN(expiresAt) || Number.isNaN(currentTime)) {
    throw new PayloadValidationError("Invalid timestamp in payload");
  }

  if (currentTime < issuedAt) {
    throw new PayloadValidationError("Payload is not yet valid");
  }

  if (currentTime > expiresAt) {
    throw new PayloadValidationError("Payload has expired");
  }
}

export function canonicalPayloadBody(payload) {
  return canonicalize(stripSignature(validatePayload(payload)));
}

export function verifySignedPayload(payload, options = {}) {
  const verifiedPayload = validatePayload(payload);
  verifyTimeWindow(verifiedPayload, options.now ?? new Date());

  let publicKeyPem = options.publicKeyPem;
  let trustAnchorKeyId = null;

  if (!publicKeyPem) {
    const trustRegistry = options.trustRegistry;
    if (!trustRegistry) {
      throw new PayloadValidationError("Missing publicKeyPem or trustRegistry option");
    }

    const resolved = resolveTrustAnchorPublicKey(trustRegistry, verifiedPayload.issuer);
    publicKeyPem = resolved.publicKeyPem;
    trustAnchorKeyId = resolved.keyId;
  }

  const canonicalBody = canonicalize(stripSignature(verifiedPayload));
  const signature = base64UrlDecode(verifiedPayload.sig);
  const ok = verifyCanonicalPayload(canonicalBody, signature, publicKeyPem, verifiedPayload.alg);

  if (!ok) {
    throw new SignatureVerificationError("Signature verification failed");
  }

  return {
    ok: true,
    issuerId: verifiedPayload.issuer.issuer_id,
    trustAnchorId: verifiedPayload.issuer.trust_anchor_id,
    trustAnchorKeyId,
    documentId: verifiedPayload.document.document_id,
    algorithm: verifiedPayload.alg
  };
}

export function verifyTransportPayload(transportPayload, options = {}) {
  const payloadJson = decodeTransportPayload(transportPayload);
  const payload = JSON.parse(payloadJson);
  return verifySignedPayload(payload, options);
}

export function validatePaymentProfile(payload, expected = {}) {
  const verifiedPayload = validatePaymentProfilePayload(payload);
  const document = verifiedPayload.document;

  const requiredFields = [
    "document_id",
    "document_type",
    "beneficiary_name",
    "iban",
    "amount",
    "currency",
    "reference",
    "due_date",
    "transaction_id",
    "communication"
  ];

  const missingFields = requiredFields.filter((field) => !validateDocumentPaymentField(document, field));
  const mismatches = [];

  if (verifiedPayload.intent !== "payment") {
    mismatches.push("intent");
  }

  if (document.document_type !== "invoice") {
    mismatches.push("document_type");
  }

  const comparisons = [
    ["document_type", expected.document_type],
    ["beneficiary_name", expected.beneficiary_name],
    ["iban", expected.iban],
    ["amount", expected.amount],
    ["currency", expected.currency],
    ["reference", expected.reference],
    ["due_date", expected.due_date],
    ["transaction_id", expected.transaction_id],
    ["communication", expected.communication]
  ];

  for (const [field, expectedValue] of comparisons) {
    if (expectedValue !== undefined && !compareExpectedValue(document[field], expectedValue)) {
      mismatches.push(field);
    }
  }

  const ok = missingFields.length === 0 && mismatches.length === 0;

  return {
    ok,
    profile: "payment.invoice",
    missingFields,
    mismatches
  };
}

File: sdk-python/document_trust_payload/verify.py
import json
from datetime import datetime, timezone

from .canonicalize import canonicalize
from .crypto import base64url_decode, verify_canonical_payload
from .errors import PayloadValidationError, SignatureVerificationError
from .qr import decode_transport_payload
from .trust_registry import resolve_trust_anchor_public_key


def _validate_required_string(value, name):
    if not isinstance(value, str) or not value:
        raise PayloadValidationError(f"Missing or invalid {name}")


def _validate_payload(payload):
    if not isinstance(payload, dict):
        raise PayloadValidationError("Payload must be a JSON object")

    _validate_required_string(payload.get("version"), "version")
    _validate_required_string(payload.get("intent"), "intent")
    _validate_required_string(payload.get("issued_at"), "issued_at")
    _validate_required_string(payload.get("expires_at"), "expires_at")
    _validate_required_string(payload.get("nonce"), "nonce")
    _validate_required_string(payload.get("alg"), "alg")
    _validate_required_string(payload.get("sig"), "sig")

    if payload["version"] != "1":
        raise PayloadValidationError(f"Unsupported version: {payload['version']}")

    issuer = payload.get("issuer")
    if not isinstance(issuer, dict):
        raise PayloadValidationError("Missing or invalid issuer object")

    _validate_required_string(issuer.get("issuer_id"), "issuer.issuer_id")
    _validate_required_string(issuer.get("display_name"), "issuer.display_name")
    _validate_required_string(issuer.get("trust_anchor_id"), "issuer.trust_anchor_id")

    document = payload.get("document")
    if not isinstance(document, dict):
        raise PayloadValidationError("Missing or invalid document object")

    _validate_required_string(document.get("document_id"), "document.document_id")
    return payload


def _validate_payment_profile_payload(payload):
    if not isinstance(payload, dict):
        raise PayloadValidationError("Payload must be a JSON object")

    _validate_required_string(payload.get("version"), "version")
    _validate_required_string(payload.get("intent"), "intent")
    _validate_required_string(payload.get("issued_at"), "issued_at")
    _validate_required_string(payload.get("expires_at"), "expires_at")
    _validate_required_string(payload.get("nonce"), "nonce")
    _validate_required_string(payload.get("alg"), "alg")

    if payload["version"] != "1":
        raise PayloadValidationError(f"Unsupported version: {payload['version']}")

    issuer = payload.get("issuer")
    if not isinstance(issuer, dict):
        raise PayloadValidationError("Missing or invalid issuer object")

    _validate_required_string(issuer.get("issuer_id"), "issuer.issuer_id")
    _validate_required_string(issuer.get("display_name"), "issuer.display_name")
    _validate_required_string(issuer.get("trust_anchor_id"), "issuer.trust_anchor_id")

    document = payload.get("document")
    if not isinstance(document, dict):
        raise PayloadValidationError("Missing or invalid document object")

    _validate_required_string(document.get("document_id"), "document.document_id")
    return payload


def _strip_signature(payload):
    body = dict(payload)
    body.pop("sig", None)
    return body


def _parse_time(value):
    parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _verify_time_window(payload, now):
    issued_at = _parse_time(payload["issued_at"])
    expires_at = _parse_time(payload["expires_at"])
    current_time = now if isinstance(now, datetime) else _parse_time(now)

    if current_time < issued_at:
        raise PayloadValidationError("Payload is not yet valid")
    if current_time > expires_at:
        raise PayloadValidationError("Payload has expired")


def canonical_payload_body(payload):
    return canonicalize(_strip_signature(_validate_payload(payload)))


def verify_signed_payload(payload, public_key_pem: str | None = None, trust_registry=None, now=None):
    verified_payload = _validate_payload(payload)
    _verify_time_window(verified_payload, now or datetime.now(timezone.utc))

    resolved_key_id = None
    if public_key_pem is None:
        if trust_registry is None:
            raise PayloadValidationError("Missing public_key_pem or trust_registry")
        resolved = resolve_trust_anchor_public_key(trust_registry, verified_payload["issuer"])
        public_key_pem = resolved["public_key_pem"]
        resolved_key_id = resolved.get("key_id")

    canonical_body = canonicalize(_strip_signature(verified_payload)).encode("utf-8")
    signature = base64url_decode(verified_payload["sig"])
    ok = verify_canonical_payload(canonical_body, signature, public_key_pem, verified_payload["alg"])

    if not ok:
        raise SignatureVerificationError("Signature verification failed")

    return {
        "ok": True,
        "issuer_id": verified_payload["issuer"]["issuer_id"],
        "trust_anchor_id": verified_payload["issuer"]["trust_anchor_id"],
        "trust_anchor_key_id": resolved_key_id,
        "document_id": verified_payload["document"]["document_id"],
        "algorithm": verified_payload["alg"],
    }


def verify_transport_payload(transport_payload, public_key_pem: str | None = None, trust_registry=None, now=None):
    payload_json = decode_transport_payload(transport_payload)
    payload = json.loads(payload_json)
    return verify_signed_payload(payload, public_key_pem=public_key_pem, trust_registry=trust_registry, now=now)


def _present(value):
    return value is not None and value != ""


def _compare_expected(actual, expected):
    if expected is None:
        return True
    if isinstance(actual, (int, float)) or isinstance(expected, (int, float)):
        return float(actual) == float(expected)
    return str(actual) == str(expected)


def validate_payment_profile(payload, expected=None):
    verified_payload = _validate_payment_profile_payload(payload)
    document = verified_payload["document"]
    expected = expected or {}

    required_fields = [
        "document_id",
        "document_type",
        "beneficiary_name",
        "iban",
        "amount",
        "currency",
        "reference",
        "due_date",
        "transaction_id",
        "communication",
    ]

    missing_fields = [field for field in required_fields if not _present(document.get(field))]
    mismatches = []

    if verified_payload.get("intent") != "payment":
        mismatches.append("intent")

    if document.get("document_type") != "invoice":
        mismatches.append("document_type")

    for field in [
        "document_type",
        "beneficiary_name",
        "iban",
        "amount",
        "currency",
        "reference",
        "due_date",
        "transaction_id",
        "communication",
    ]:
        if field in expected and not _compare_expected(document.get(field), expected.get(field)):
            mismatches.append(field)

    return {
        "ok": not missing_fields and not mismatches,
        "profile": "payment.invoice",
        "missing_fields": missing_fields,
        "mismatches": mismatches,
    }
